src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp

```diff
@@ -16,6 +17,99 @@
 namespace stan {
 namespace analyze {
 
+/**
+ * Computes normalized average ranks for draws. Transforming them to normal
+ * scores using inverse normal transformation and a fractional offset. Based on
+ * paper https://arxiv.org/abs/1903.08008
+ * @param draws stores chains in columns
+ * @return normal scores for average ranks of draws
+ *
+ */
+
+Eigen::MatrixXd rank_transform(const Eigen::MatrixXd& draws) {
+  int rows = draws.rows();
+  int cols = draws.cols();
+  int size = rows * cols;
+  Eigen::MatrixXd rankMatrix = Eigen::MatrixXd::Zero(rows, cols);
+
+  std::vector<std::pair<double, int>> valueWithIndex(size);
+
+  for (int col = 0; col < cols; ++col) {
+    for (int row = 0; row < rows; ++row) {
+      int index
+          = col * rows + row;  // Calculating linear index in column-major order
+      valueWithIndex[index] = {draws(row, col), index};
+    }
+  }
+
+  std::sort(valueWithIndex.begin(), valueWithIndex.end());
+
+  // Assigning average ranks
+  for (int i = 0; i < size; ++i) {
+    // Handle ties by averaging ranks
+    int j = i;
+    double sumRanks = 0;
+    int count = 0;
+
+    while (j < size && valueWithIndex[j].first == valueWithIndex[i].first) {
+      sumRanks += j + 1;  // Rank starts from 1
+      ++j;
+      ++count;
+    }
```

Comment:
Right now this while loop while always go off at least once, can you start `j` at `j = i + 1` etc to avoid this? Then that while loop only happens for cases of duplicates